### PR TITLE
[Humble] Fix split by time. (backport #1022)

### LINKED ISSUE #1602: wrong duration and missing topics/frames after ros2 bag convert
## Description
I have a bagfile and wanted to split it into several files using `ros2 bag convert`, but the resulting bagfile is not working as expected.

### Expected Behavior
I have the original bagfile with a single file of nearly 4GB. With convert I want to split this file into ~500MB parts.
Both, the old and the new bagfile, should behave the same.

### Actual Behavior
When I use `ros2 bag play` on the old bagfile, everything works fine. But when using the new bagfile, i get a lot missing frames and topics.
When inspecting both files with `ros2 bag info` the topic information and message count is the same, but the duration and start time differs.

## To Reproduce
** Steps to reproduce the behavior, e.g.
1. Run `ros2 bag play rosbag_origin` to validate you have a working bagfile. No erros in the logs and the process in running ~80s.
2. Run `ros2 bag convert` to modify the bagfile:
```
ros2 bag convert -i rosbag_origin -o convert.yml
[INFO] [1712645045.805924828] [rosbag2_storage]: Opened database 'rosbag_origin/rosbag_origin_0.db3' for READ_ONLY.
[INFO] [1712645045.810819109] [rosbag2_storage]: Opened database 'rosbag_convert/rosbag_convert_0.db3' for READ_WRITE.
[INFO] [1712645047.498534912] [rosbag2_storage]: Opened database 'rosbag_convert/rosbag_convert_1.db3' for READ_WRITE.
[INFO] [1712645049.129373085] [rosbag2_storage]: Opened database 'rosbag_convert/rosbag_convert_2.db3' for READ_WRITE.
[INFO] [1712645050.516206986] [rosbag2_storage]: Opened database 'rosbag_convert/rosbag_convert_3.db3' for READ_WRITE.
[INFO] [1712645051.811610194] [rosbag2_storage]: Opened database 'rosbag_convert/rosbag_convert_4.db3' for READ_WRITE.
[INFO] [1712645053.199245268] [rosbag2_storage]: Opened database 'rosbag_convert/rosbag_convert_5.db3' for READ_WRITE.
[INFO] [1712645054.700044665] [rosbag2_storage]: Opened database 'rosbag_convert/rosbag_convert_6.db3' for READ_WRITE.
[INFO] [1712645056.326892823] [rosbag2_storage]: Opened database 'rosbag_convert/rosbag_convert_7.db3' for READ_WRITE.
[INFO] [1712645058.296396332] [rosbag2_storage]: Opened database 'rosbag_convert/rosbag_convert_8.db3' for READ_WRITE.
```
with convert.yml
```
output_bags:
  - uri: rosbag_convert
    all: true
    max_bagfile_size: 500000000
```
3. Run `ros2 bag play rosbag_convert`, which will end after ~9s. Furthermore, the individual files open within milliseconds.
4. Run `ros2 bag info` on both files to see different durations and times:
```
ros2 bag info rosbag_origin
  Files:             rosbag_origin_0.db3
  Bag size:          3.8 GiB
  Storage id:        sqlite3
  Duration:          80.229s
  Start:             Apr  8 2024 10:47:45.72 (1712566065.72)
  End:               Apr  8 2024 10:49:05.301 (1712566145.301)
  Messages:          49329
```
```
ros2 bag info rosbag_convert
  Files:             rosbag_convert_0.db3
                     rosbag_convert_1.db3
                     rosbag_convert_2.db3
                     rosbag_convert_3.db3
                     rosbag_convert_4.db3
                     rosbag_convert_5.db3
                     rosbag_convert_6.db3
                     rosbag_convert_7.db3
                     rosbag_convert_8.db3
  Bag size:          3.8 GiB
  Storage id:        sqlite3
  Duration:          9.895s
  Start:             Apr  8 2024 10:49:03.894 (1712566143.894)
  End:               Apr  8 2024 10:49:13.790 (1712566153.790)
  Messages:          49329
```

## System (please complete the following information)
 - OS: Ubuntu 22.04
 - ROS 2 Distro: Humble
 - Install Method: APT
 - Version: 0.15.9

## Additional context
I've changed the file names in the output - if there are inconsistent names in the output, I'm sorry.
I used `apt upgrade` to install the latest ros2bag version

### PR DESCRIPTION
**Pull request by fujitatomoya**
humble backport of https://github.com/ros2/rosbag2/pull/1022

### CODE REVIEW

File: rosbag2_cpp/src/rosbag2_cpp/writers/sequential_writer.cpp [hunk 1]
```diff
@@ -305,10 +307,14 @@ void SequentialWriter::write(std::shared_ptr<rosbag2_storage::SerializedBagMessa
   const auto message_timestamp = std::chrono::time_point<std::chrono::high_resolution_clock>(
     std::chrono::nanoseconds(message->time_stamp));
 
-  if (should_split_bagfile(message_timestamp)) {
-    split_bagfile();
+  if (is_first_message_) {
```
Comment (fujitatomoya) [inline rosbag2_cpp/src/rosbag2_cpp/writers/sequential_writer.cpp:310]: `is_first_message_` flag already exists (came in with https://github.com/ros2/rosbag2/pull/1037), so we will not break ABI compatibility at all.

`colcon test --event-handlers console_direct+ --packages-select rosbag2_cpp` test all passes.
i verified that [original issue](https://github.com/ros2/rosbag2/issues/966) still exists in humble, but this patch can resolve that.

**Comment by fujitatomoya** [comment]
@Kaju-Bubanja @MichaelOrlov i could use review on this. (this is good for humble user.)

**Comment by Barry-Xu-2018** [comment]
LGTM

**Comment by Kaju-Bubanja** [comment]
lgtm

**Comment by fujitatomoya** [comment]
@MichaelOrlov i will leave this to you. this obviously fixes the critical problem for humble users without breaking ABI.

**Review by MichaelOrlov — CHANGES_REQUESTED** [review]
@fujitatomoya I don't mind about this PR if it is fixing problem and not breaking ABI/API.

I found that one line change from the original PR is missing
in line 
https://github.com/ros2/rosbag2/blob/04afde654f153a02539f83bcaac75811f3e83e0f/rosbag2_cpp/src/rosbag2_cpp/writers/sequential_writer.cpp#L319

was changed  to the 
```cpp
const auto duration = message_timestamp - metadata_.files.back().starting_time;
```
If it wasn't intentionally skipped  please add those changes.

**Comment by MichaelOrlov** [comment]
@fujitatomoya, Could you please also take a look at why `ros2bag` tests fail? It seems to be something related to the Python imports.

**Comment by fujitatomoya** [comment]
> @fujitatomoya I don't mind about this PR if it is fixing problem and not breaking ABI/API.
> 
> I found that one line change from the original PR is missing in line
> 
> https://github.com/ros2/rosbag2/blob/04afde654f153a02539f83bcaac75811f3e83e0f/rosbag2_cpp/src/rosbag2_cpp/writers/sequential_writer.cpp#L319
> 
> was changed to the
> 
> ```c++
> const auto duration = message_timestamp - metadata_.files.back().starting_time;
> ```
> 
> If it wasn't intentionally skipped please add those changes.

@MichaelOrlov we cannot take that line. this has been addressed by https://github.com/ros2/rosbag2/pull/1098, otherwise `split_bag_metadata_has_full_duration` fails.

**Comment by fujitatomoya** [comment]
@MichaelOrlov can we know which test failed, the log is really long. could not find the specific test failure. if that is the python import, that would be unrelated.

**Comment by MichaelOrlov** [comment]
@fujitatomoya It fails with the following output
<details><summary>Click to expand</summary>

```
2024-04-17T19:32:57.3786356Z Starting >>> ros2bag
2024-04-17T19:32:58.5114970Z --- output: ros2bag
2024-04-17T19:32:58.5115445Z ============================= test session starts ==============================
2024-04-17T19:32:58.5116031Z platform linux -- Python 3.10.12, pytest-8.1.1, pluggy-1.4.0
2024-04-17T19:32:58.5116569Z cachedir: /__w/rosbag2/rosbag2/ros_ws/build/ros2bag/.pytest_cache
2024-04-17T19:32:58.5117163Z rootdir: /__w/rosbag2/rosbag2/ros_ws/src/ww3gvnpn3ab/rosbag2/ros2bag
2024-04-17T19:32:58.5117595Z configfile: pytest.ini
2024-04-17T19:32:58.5119222Z plugins: launch_testing_ros-0.19.7, launch_testing-1.0.5, ament_copyright-0.12.10, ament_pep257-0.12.10, ament_flake8-0.12.10, ament_lint-0.12.10, mock-3.14.0, colcon-core-0.15.1, rerunfailures-14.0, repeat-0.9.3, cov-5.0.0, timeout-2.1.0
2024-04-17T19:32:58.5120465Z collected 0 items / 1 error
2024-04-17T19:32:58.5120655Z 
2024-04-17T19:32:58.5120809Z ==================================== ERRORS ====================================
2024-04-17T19:32:58.5121272Z ____________________________ ERROR collecting test _____________________________
2024-04-17T19:32:58.5121915Z /usr/local/lib/python3.10/dist-packages/pluggy/_hooks.py:501: in __call__
2024-04-17T19:32:58.5122557Z     return self._hookexec(self.name, self._hookimpls.copy(), kwargs, firstresult)
2024-04-17T19:32:58.5123263Z /usr/local/lib/python3.10/dist-packages/pluggy/_manager.py:119: in _hookexec
2024-04-17T19:32:58.5123872Z     return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
2024-04-17T19:32:58.5124579Z /usr/local/lib/python3.10/dist-packages/_pytest/python.py:225: in pytest_collect_file
2024-04-17T19:32:58.5125267Z     module: Module = ihook.pytest_pycollect_makemodule(
2024-04-17T19:32:58.5125896Z /usr/local/lib/python3.10/dist-packages/_pytest/config/compat.py:79: in fixed_hook
2024-04-17T19:32:58.5126396Z     return hook(**kw)
2024-04-17T19:32:58.5126851Z /usr/local/lib/python3.10/dist-packages/pluggy/_hooks.py:501: in __call__
2024-04-17T19:32:58.5127467Z     return self._hookexec(self.name, self._hookimpls.copy(), kwargs, firstresult)
2024-04-17T19:32:58.5128157Z /usr/local/lib/python3.10/dist-packages/pluggy/_manager.py:119: in _hookexec
2024-04-17T19:32:58.5128757Z     return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
2024-04-17T19:32:58.5129416Z ../../../../build/launch_testing/launch_testing/pytest/hooks.py:188: in pytest_pycollect_makemodule
2024-04-17T19:32:58.5130003Z     entrypoint = find_launch_test_entrypoint(path)
2024-04-17T19:32:58.5130584Z ../../../../build/launch_testing/launch_testing/pytest/hooks.py:178: in find_launch_test_entrypoint
2024-04-17T19:32:58.5131138Z     module = import_path(path, root=None)
2024-04-17T19:32:58.5131781Z E   TypeError: import_path() missing 1 required keyword-only argument: 'consider_namespace_packages'
2024-04-17T19:32:58.5132558Z --- generated xml file: /__w/rosbag2/rosbag2/ros_ws/build/ros2bag/pytest.xml ---
2024-04-17T19:32:58.5133097Z =========================== short test summary info ============================
2024-04-17T19:32:58.5133831Z ERROR test - TypeError: import_path() missing 1 required keyword-only argument: 'consider_namespace_packages'
2024-04-17T19:32:58.5134529Z !!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!
2024-04-17T19:32:58.5134987Z =============================== 1 error in 0.13s ===============================
2024-04-17T19:32:58.5135345Z ---
2024-04-17T19:32:58.5135589Z Failed   <<< ros2bag [1.17s, exited with code 2]
```
</details>

**Review by MichaelOrlov — APPROVED** [review]
@fujitatomoya LGTM with green CI.
Thanks for the explanation about missed delta for "duration" variable, and reference to the https://github.com/ros2/rosbag2/pull/1098. Make sense to me.
As regards to the `ros2bag` failure.  
I've tried to checkout your branch and build and test it on my local environment. The `ros2bag` package in particularly compiles and passes tests without errors. 
Approving then.

**Comment by fujitatomoya** [comment]
CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/20915/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/15216/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/21593/)

**Comment by fujitatomoya** [comment]
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/21596/)

**Comment by fujitatomoya** [comment]
https://ci.ros2.org/job/ci_windows/21596/ is unrelated.

@MichaelOrlov can you merge this backport PR to humble?

**Pull request merged**